diag_manager/fms_diag_yaml.F90

```diff
@@ -1,62 +1,328 @@
+  if (natt .ne. 0) then
+    call get_block_ids(file_id, "attributes", var_att_id, parent_block_id=var_id)
+    nkeys = get_nkeys(file_id, var_att_id(1))
+    allocate(key_ids(nkeys))
+    call get_key_ids(file_id, var_att_id(1), key_ids)
+
+    allocate(field%var_attributes(nkeys, 2))
+    do j = 1, nkeys
+      call get_key_name(file_id,  key_ids(j), field%var_attributes(j, 1))
+      call get_key_value(file_id, key_ids(j), field%var_attributes(j, 2))
+    enddo
+    deallocate(key_ids)
+  endif
+
+end subroutine
+
+!< @brief diag_manager wrapper to get_value_from_key to use for allocatable
+!! string variables
+subroutine diag_get_value_from_key(file_id, par_id, key_name, value_name, is_optional)
+  integer,           intent(in) :: file_id     !< Id of the diag_table yaml file
+  integer,           intent(in) :: par_id      !< Id of the parent block in the yaml file
+  character(len=*),  intent(in) :: key_name    !< Key to look for in the parent block
+  character(len=:), allocatable :: value_name  !< Value of the key
+  logical, intent(in), optional :: is_optional !< Flag indicating if the key is optional
+
+  character(len=255) :: buffer !< String buffer to read in to
+
+  call get_value_from_key(file_id, par_id, trim(key_name), buffer, is_optional= is_optional)
+  allocate(character(len=len_trim(buffer)) :: value_name)
+  value_name = trim(buffer)
+
+end subroutine diag_get_value_from_key
+
+!< @brief gets the lat/lon of the sub region to use in a diag_table yaml
+subroutine get_sub_region_latlon(diag_yaml_id, sub_region_id, lat_lon_sub_region)
+  integer, intent(in)  :: diag_yaml_id       !< Id of the diag_table yaml file
+  integer, intent(in)  :: sub_region_id      !< Id of the region block to read from
+  real,    intent(out) :: lat_lon_sub_region (NUM_SUB_REGION_ARRAY) !< Array storing the bounds of the sub region
+
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim1_begin", lat_lon_sub_region(1), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim1_end", lat_lon_sub_region(2), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim2_begin", lat_lon_sub_region(3), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim2_end", lat_lon_sub_region(4), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim3_begin", lat_lon_sub_region(5), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim3_end", lat_lon_sub_region(6), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim4_begin", lat_lon_sub_region(7), is_optional=.true.)
+  call get_value_from_key(diag_yaml_id, sub_region_id, "dim4_end", lat_lon_sub_region(8), is_optional=.true.)
+
+end subroutine get_sub_region_latlon
+
+!< @brief gets the indexes of the sub region to use in a diag_table yaml
+subroutine get_sub_region_index(diag_yaml_id, sub_region_id, index_sub_region)
```

Comment:
There's gotta be a cleaner way to do this than have 2 identical subroutines but one is for real and one is for integer.  Can you have the `out` variable be `class(*)` and let `get_value_from_key` figure it out?  Is that subroutine polymorphic?

Reply:
https://github.com/NOAA-GFDL/FMS/pull/866/commits/c56ebeedcec7cb0eaa022248af71d77305585a80

Reply:
This is much better!